libyul/YulName.h

```diff
@@ -0,0 +1,230 @@
+	YulName nameOfLabel(std::string_view label) const;
+
+	/// Same as nameOfLabel, but restricted to builtins and therefore computationally more efficient.
+	YulName nameOfBuiltin(std::string_view builtin) const;
+
+	/// Same as nameOfLabel, but restricted to types and therefore computationally more efficient.
+	YulName nameOfType(std::string_view type) const;
+
+	/// Whether the name is corresponding to a type.
+	[[nodiscard]] bool isType(YulName _name) const;
+
+	/// Number of types. If the dialect is untyped, there is still one type (the "empty type" type)
+	size_t nTypes() const;
+
+	/// The contained dialect.
+	Dialect const& dialect() const;
+
+	/// Whether the contained dialect is an EVM dialect (see ``EVMDialect``).
+	[[nodiscard]] bool isEvmDialect() const;
+
+	/// Generates labels for derived names over the set of _usedNames, respecting a set of _illegal labels.
+	/// This will change the state of all derived names in _usedNames to "not derived" with a label associated to them.
+	void generateLabels(std::set<YulName> const& _usedNames, std::set<std::string> const& _illegal = {});
+
+	// commented out for the time being, as it requires AST refactoring to be YulName-based
+	// void generateLabels(Block const& _ast, std::set<std::string> const& _illegal = {});
+
+private:
+	struct PredefinedBuiltinFunctions
+	{
+		std::vector<std::optional<YulName>> discardFunctions;
+		std::vector<std::optional<YulName>> equalityFunctions;
+		std::optional<YulName> booleanNegationFunction;
+		std::vector<std::optional<YulName>> memoryStoreFunctions;
+		std::vector<std::optional<YulName>> memoryLoadFunctions;
+		std::vector<std::optional<YulName>> storageStoreFunctions;
+		std::vector<std::optional<YulName>> storageLoadFunctions;
+		std::vector<YulName> hashFunctions;
+	};
+	struct IndexBoundaries
+	{
+		size_t beginTypes {};
+		size_t endTypes {};
+		size_t beginBuiltins {};
+		size_t endBuiltins {};
+	};
+	enum class YulNameState	{ DERIVED, DEFINED };
+	bool nameWithinBounds(YulName const _name) const { return _name < m_index; }
+
+	size_t indexOfType(YulName _type) const;
+	BuiltinFunction convertBuiltinFunction(YulName _name, yul::BuiltinFunction const& _builtin) const;
+	BuiltinFunction const* fetchTypedPredefinedFunction(YulName _type, std::vector<std::optional<YulName>> const& _functions) const;
+
+	std::reference_wrapper<Dialect const> m_dialect;
+	std::vector<std::tuple<YulName, std::string>> m_dialectTypes;
+	std::map<YulName, BuiltinFunction> m_builtinFunctions;
+
+	PredefinedBuiltinFunctions m_predefinedBuiltinFunctions;
+
+	size_t m_index {0};
+	size_t m_nGhosts {0};
+	std::vector<std::string> m_definedLabels {};
+	std::vector<std::tuple<YulName, YulNameState>> m_names {};
+	std::map<std::tuple<size_t, size_t>, YulName> m_verbatimNames {};
+	PredefinedHandles m_predefined{};
+	IndexBoundaries m_indexBoundaries;
+};
+using YulName = YulNameRepository::YulName;
```

Comment:
It's odd that this does not break anything given that after #15083 `_yulname` is producing `YulString`s while here we change the underlying type of most things that those `YulString`s get assigned to.

Reply:
yeah, there was that. rebased. :)